distribution: errors: do not retry if no credentials provided

runcom opened the pull request:
Fixes case `c)` in #21054

/cc @aaronlehmann 

Signed-off-by: Antonio Murdaca runcom@redhat.com

runcom committed:
distribution: errors: do not retry if no credentials provided

Fix and add test for case c) in #21054

Signed-off-by: Antonio Murdaca <runcom@redhat.com>

runcom commented:
This should probably differentiate between pull and push I guess

aaronlehmann commented:
Thanks for the PR!

aaronlehmann commented on integration-cli/docker_cli_push_test.go:
```diff
@@ -526,3 +526,11 @@ func (s *DockerTrustSuite) TestTrustedPushWithReleasesDelegation(c *check.C) {
+	repoName := fmt.Sprintf("%s/busybox", privateRegistryURL)
+	dockerCmd(c, "tag", "busybox", repoName)
+	out, err := dockerCmdWithError("push", repoName)
+	c.Assert(err, check.NotNil, check.Commentf(out))
```
Should this check for a specific error message?

Maybe we should check the output to make sure it doesn't include `Retrying`?

runcom replied:
I'm having troubles running tests on my laptop so I did a test PR while fixing this test

aaronlehmann commented:
Weird, CI didn't run on this PR.

ping @icecrime

thaJeztah commented:
@aaronlehmann oh, that happens sometimes if GitHub misses a ping. Doing a `!rebuild` will help, let me do that

thaJeztah commented:
Yup, worked, they're here now :-)

runcom commented:
Btw I think this fix will regress on the pull side, it won't return a 404 anymore when credentials aren't found I suppose (need to check)

runcom committed:
integration-cli: add tests for case a) and d) in #21054

- add test for pull from private registry with no credentials
- add test for push to docker hub with no credentials

Signed-off-by: Antonio Murdaca <runcom@redhat.com>
Signed-off-by: Antonio Murdaca <amurdaca@localhost.localdomain>
Signed-off-by: Antonio Murdaca <runcom@redhat.com>

runcom commented:
This PR is also adding tests for case a) and d) - we'll reiterate after we *fix* everything in #21054

runcom commented:
@aaronlehmann this is good to review if all tests pass :)

runcom commented on integration-cli/docker_cli_push_test.go:
```diff
@@ -526,3 +526,21 @@ func (s *DockerTrustSuite) TestTrustedPushWithReleasesDelegation(c *check.C) {
+	out, _, err := dockerCmdWithError("push", repoName)
+	c.Assert(err, check.NotNil, check.Commentf(out))
+	c.Assert(out, check.Not(checker.Contains), "Retrying")
+	c.Assert(out, checker.Contains, "no basic auth credentials")
```
I'm wondering if the error string here should be the same as the docker.io case below

runcom commented on integration-cli/docker_cli_push_test.go:
```diff
@@ -526,3 +526,21 @@ func (s *DockerTrustSuite) TestTrustedPushWithReleasesDelegation(c *check.C) {
+
+// This may be flaky but it's needed not to regress on unauthorized push, see #21054
+func (s *DockerSuite) TestPushToCentralRegistryUnauthorized(c *check.C) {
+	repoName := "test/busybox"
```
This needs testrequires network (I'll add it tomorrow)

calavera commented:
LGTM

aaronlehmann commented:
LGTM
